### PR TITLE
BlockTags should not be reused if they have changed in any way that their consumer might be able to observe.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/14087
Fixes https://github.com/dotnet/roslyn/issues/14123

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide

File: src/EditorFeatures/Next/Structure/VisualStudio15StructureTaggerProvider.cs [hunk 1]
```diff
@@ -85,6 +74,36 @@ private class RoslynBlockTag : RoslynOutliningRegionTag, IBlockTag
...
+                       this.StatementSpan == tag.StatementSpan &&
+                       this.Span == tag.Span;
+            }
+
+            public override int GetHashCode()
+            {
+                return Hash.Combine(this.Level,
+                       Hash.Combine(this.Type, 
+                       Hash.Combine(this.CollapsedForm,
+                       Hash.Combine(this.StatementSpan.GetHashCode(), this.Span.GetHashCode()))));
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Next/Structure/VisualStudio15StructureTaggerProvider.cs]: Why the .GetHashCode on some but not others?

File: src/EditorFeatures/Next/Structure/VisualStudio15StructureTaggerProvider.cs [hunk 2]
```diff
@@ -85,6 +74,36 @@ private class RoslynBlockTag : RoslynOutliningRegionTag, IBlockTag
...
             }
 
+            public override bool Equals(object obj)
+                => Equals(obj as RoslynBlockTag);
+
+            /// <summary>
+            /// This is only called if the spans for the tags were the same.  However, even if we 
+            /// have the same span as the previous tag (taking into account span mapping) that 
+            /// doesn't mean we can use the old block tag.  Specifically, the editor will look at
+            /// other fields in the tags So we need to make sure that these values have not changed
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Next/Structure/VisualStudio15StructureTaggerProvider.cs]: Missing period.

File: src/EditorFeatures/Next/Structure/VisualStudio15StructureTaggerProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Next/Structure/VisualStudio15StructureTaggerProvider.cs]: Unintentional.  Will update in my followup change.  Thanks!

**Pull request merged**